### PR TITLE
Shorten lifetime of panic temporaries in panic_fmt case

### PR DESCRIPTION
**Pull request by dtolnay**
This fixes an issue called out by @fasterthanlime in https://octodon.social/@fasterthanlime/109304454114856561. Macros like `todo!("…")` and `panic!("…", …)` drop their `format_args` temporary at the nearest enclosing semicolon **outside** the macro invocation, not inside the macro invocation. Due to the formatting internals being type-erased in a way that is not thread-safe, this blocks futures from being `Send` if there is an `await` anywhere between the panic call and the nearest enclosing semicolon.

**Example:**

```rust
#![allow(unreachable_code)]
 
async fn f(_: u8) {}
 
async fn g() {
    f(todo!("...")).await;
}
 
fn require_send(_: impl Send) {}
 
fn main() {
    require_send(g());
}
```

**Before:**

```console
error: future cannot be sent between threads safely
  --> src/main.rs:15:18
   |
15 |     require_send(g());
   |                  ^^^ future returned by `g` is not `Send`
   |
   = help: the trait `Sync` is not implemented for `core::fmt::Opaque`
note: future is not `Send` as this value is used across an await
  --> src/main.rs:9:20
   |
9  |     f(todo!("...")).await;
   |       ------------ ^^^^^^ await occurs here, with `$crate::format_args!($($arg)+)` maybe used later
   |       |
   |       has type `ArgumentV1<'_>` which is not `Send`
note: `$crate::format_args!($($arg)+)` is later dropped here
  --> src/main.rs:9:26
   |
9  |     f(todo!("...")).await;
   |                          ^
note: required by a bound in `require_send`
  --> src/main.rs:12:25
   |
12 | fn require_send(_: impl Send) {}
   |                         ^^^^ required by this bound in `require_send`
```

**After:** works.

Arguably there is a rustc fix that could work here too, instead of a standard library change. Rustc could be taught that the code shown above is fine to compile because the `await` is unreachable and so temporaries before the `await` do not need to get put in the anonymous compiler-generated `Future` struct, regardless of syntactically where they're supposed to be dropped according to the language semantics.

I would be open to that, though my recollection is that in the past we have been very hesitant about introducing any smarts into Drop placement. People want the language rules about where temporaries are dropped to be as simplistic and predictable as possible.

### CODE REVIEW

**Comment by Aaron1011** [comment]
This seems very similar to https://github.com/rust-lang/rust/pull/64856, but I'm not sure if this case has observable drop-order effects.

**Comment by dtolnay** [comment]
Maybe I should know better after #99689, but I was not able to come up with any program where this PR would cause a change in drop order at runtime.

**Comment by joshtriplett** [comment]
This seems like a reasonable change to me. I'd want to see it pass a crater run (either directly or as part of the normal beta crater runs) but otherwise :+1:.

**Comment by dtolnay** [comment]
Beta craters are going to be sufficient here I think.

**Comment by dtolnay** [comment]
- Rebased to resolve conflict with #106458

**Comment by dtolnay** [comment]
- Rebased to resolve conflict with #106745

**Comment by joshtriplett** [comment]
@bors r+

If any beta crater issues point to this it can be reverted.

**Comment by dtolnay** [comment]
The failure in src/tools/clippy/tests/ui/diverging_sub_expression.stderr is related to https://github.com/rust-lang/rust-clippy/issues/10776. I think this is okay to land without fixing that in clippy first.

File: library/core/src/panic.rs [hunk 1]
```diff
@@ -53,9 +55,11 @@ pub macro panic_2021 {
     ("{}", $arg:expr $(,)?) => (
         $crate::panicking::panic_display(&$arg)
```
Comment (bjorn3) [inline library/core/src/panic.rs:56]: Should the same be done here?

Comment (dtolnay) [reply library/core/src/panic.rs:56]: Followup: #111590.

**Pull request merged**